### ISSUE TITLE
TagFirstLast is not optimal in memory

### ISSUE THREAD

**Issue by Orace** [morelinq/MoreLINQ#625]
`TagFirstLast` use `CountDown` and therefor use a `Queue` of all the elements.
Store the full sequence could and should be avoided.

**Comment by atifaziz** [comment]
> use a `Queue` of all the elements

How's that? It uses a queue that's a count of 1 in the absolute worst case:

https://github.com/morelinq/MoreLINQ/blob/5205ea241d72b079436060d330cd5c2eae7cdcdf/MoreLinq/TagFirstLast.cs#L64

In turn `CountDown` is optimised for collections and lists:

https://github.com/morelinq/MoreLINQ/blob/5205ea241d72b079436060d330cd5c2eae7cdcdf/MoreLinq/CountDown.cs#L60-L64

A queue (of 1 in case of `TagFirstLast`) is only used for a pure sequence.

`TagFirstLast` used to be [implemented in terms of a composition of many other operations but then got simplified into `Index` + `CountDown`](https://github.com/morelinq/MoreLINQ/commit/4a4767141b9f9cf3996f1ec65390f08427445608#diff-342b8cde8587717089c9d21315cc985b).

**Comment by Orace** [comment]
Sorry about that, I misunderstood CountDown, that why I commented about SkipLast and TagFirstLast. Anyway my comments on SkipLast about many object creation and manipulation work here too and I will do perf test before continue along this way.

**Comment by atifaziz** [comment]
What would help here is to optimise `CountDown` for 1 so it doesn't even have to allocate a queue in that case and would eventually help `TagFirstLast` as a result. [I did something like that for `MinBy`/`MaxBy` where I used different storage strategies based on count](https://github.com/morelinq/MoreLINQ/blob/cdc6f76f4bf348e860b9eaa072279ee2956d6de7/MoreLinq/MaxBy.cs). Optimisation of `CountDown` is a separate issue so I am going to close this one.
